### ISSUE TITLE
Is the optimal assignment (Hungarian algorithm) necessary?

### ISSUE THREAD

**Issue by Benzlxs** [facebookresearch/detr#73]
Hi,

Many thanks for your novel and interesting work, and it is really impressive.

In the paper, the matching between ground truth set and predictions are solved as an assignment problem with the Hungarian algorithm, but can we pre-define the order of predictions so the assignment task can be avoided? 
To be more specific, the ground truth set (including N bounding boxes) actually can be ordered according to some defined rules, like left-top to right-down in the image, then the re-ordered ground truth set can be assigned to predictions from 1 to N. The DETR will learn to detect objects in the order of left-top to right-down, so that we don't need to do the assignment work.

Do you try this before?

**Comment by alcinos** [comment]
Hi @Benzlxs, thanks for your interest in DETR.

We did try a few fixed orderings, including arbitrary fixed order (eg the dataset order), as well as a few lexicographical orders (sorting by position, size, class, ...), but all these attempts performed worse.

If you enforce an ordering, you are essentially putting the additional constraint that the network must output the "true" predictions in the first queries, and then all predict "no-object" for all the remaining queries. Say if there are 100 object queries, which is the default in DETR, but only 3 objects to detect, then predictions [1, 3] must match the targets (according to whichever order you chose), and predictions [4, 100] must be "no objects".
By contrast, with the Hungarian matching, it doesn't matter where the "true" predictions are, they can be scattered anywhere amongst the predictions. In my example, the network can decide to use queries say 43, 57 and 99 to predict the objects, and fill the rest with "no-object".

Here is some intuition why this works better: 

- Since the network doesn't have to push the predictions to the beginning, it can instead let each query specialize in its own kind of objects. We show in the paper that each object query tend to predict objects in a specific region of the image, which would not be possible with a forced ordering.
- Related to the previous point, it's rather clear that the fixed ordering is a worse usage of the queries. In coco for ex, there is no image with more than ~75 objects. With a fixed ordering, it means that the object queries in [75-100] will never be used. Similarly, there are few images with more than >50 objects, so queries [50-75] could potentially overfit to the said images. By contrast, with the Hungarian matching, increasing the number of queries improve the recall and thus the AP.
- Finally, Hungarian matching is more robust to noise in the annotations. Say for example there are 3 objects visible A, B and C, but for some reason only B and C are annotated. In the fixed ordering case, if the model predicts A, B, C (which would be theoretically correct), since the fixed ordering loss expects B, C, "no-object" it means that the model will be hugely penalized (none of the predictions will be correct). By contrast, with a Hungarian loss, B and C will be correctly reinforced, and the network will suffer only a small classification penalty for predicting A.

I hope this helps giving more intuition about that.
I believe I have answered your question, and as such I'm closing this issue. Feel free to reach out if you have further concerns.
